proposals/resource_cache.md

```diff
@@ -0,0 +1,171 @@
+# Meta
+[meta]: #meta
+- Name: Resource Cache
+- Start Date: Aug 7, 2023
+- Update data (optional): Aug 7, 2023
+- Author(s): @JimBugwadia
+
+# Table of Contents
+[table-of-contents]: #table-of-contents
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+- [CRD Changes (OPTIONAL)](#crd-changes-optional)
+
+# Overview
+[overview]: #overview
+
+Optional caching of any Kubernetes resource.
+
+
+# Motivation
+[motivation]: #motivation
+
+From: https://github.com/kyverno/kyverno/issues/4459
+
+> In some cases to properly validate a resource we need to examine other resources. Particularly for Ingress and Istio Gateways/VirtualServices we need to look at all the other Ingress/virtualservices or services in the cluster. At large scale we are finding that Kyverno struggles to handle repeatedly pulling thousands of resources using the context apiCall variables. On a cluster with around 4,000 Service objects and 3,000 Ingresses we found that a policy validating Istio VirtualService destinations against Services (ensuring the target exists) was taking > 10s for all measurements (p50, p95 and p99), and the webhook timeout was exceeded. Another policy that validates an Ingress doesn't duplicate a hostname from another Ingress had p95 execution times over 5 seconds. During this time the controllers were at/below requested values for cpu/memory and otherwise had no other indicator of performance problems.
+
+# Proposal
+
+There are two aspects to this feature:
+1. Allow users to manage which resources should be cached
+2. Allow policy rules to reference cached resources
+
+Users can manage which resources to cache using the same mechanism that is currently used for ConfigMap resources i.e. adding a label `cache.kyverno.io/enabled: "true"` to the resource.
+
+To reference cached resources, the `apiCall` context entry can be used:
+
+```yaml
+      context:
+        - name: hosts
+          apiCall:
+            urlPath: "/apis/networking.k8s.io/v1/ingresses"
+            jmesPath: "items[].spec.rules[].host"
+            cache: true
+```
+
+# Implementation
```

Comment:
Don't see it addressed here, but with `cache: true` will Kyverno _ONLY_ return resources specified by that apiCall AND those which have been successfully cached (have `cache.kyverno.io/enabled: "true"`)? Ex., the call for Ingress resources is used per above. 20 Ingress resources exist but 0 have been labeled. Are exactly 0 returned? If so, does the call fail because of 0 or succeed with an empty set?

What about if there are still 20 total but 10 have been labeled? Will 10 or 20 be returned?

Reply:
separated cache from API call